shell/platform/windows/accessibility_root_node.cc

```diff
@@ -0,0 +1,271 @@
+  if (V_I4(&var_id) == CHILDID_SELF) {
+    std::wstring name = L"ROOT_NODE_VIEW";
+    *name_bstr = SysAllocString(name.c_str());
+    return S_OK;
+  }
+  IAccessible* target;
+  if ((target = GetTargetAndChildID(&var_id))) {
+    return window_accessible_->get_accName(var_id, name_bstr);
+  }
+  return S_FALSE;
+}
+
+IFACEMETHODIMP AccessibilityRootNode::get_accParent(IDispatch** disp_parent) {
+  return S_FALSE;
+}
+
+IFACEMETHODIMP AccessibilityRootNode::get_accRole(VARIANT var_id,
+                                                  VARIANT* role) {
+  IAccessible* target;
+  if ((target = GetTargetAndChildID(&var_id))) {
+    return target->get_accRole(var_id, role);
+  }
+  return E_FAIL;
+}
+
+IFACEMETHODIMP AccessibilityRootNode::get_accState(VARIANT var_id,
+                                                   VARIANT* state) {
+  IAccessible* target;
+  if ((target = GetTargetAndChildID(&var_id))) {
+    return target->get_accState(var_id, state);
+  }
+  return E_FAIL;
+}
+
+IFACEMETHODIMP AccessibilityRootNode::get_accHelp(VARIANT var_id, BSTR* help) {
+  if (!help) {
+    return E_INVALIDARG;
+  }
+  *help = {};
+  return S_FALSE;
+}
+
+IFACEMETHODIMP AccessibilityRootNode::get_accValue(VARIANT var_id,
+                                                   BSTR* value) {
+  IAccessible* target;
+  if ((target = GetTargetAndChildID(&var_id))) {
+    return target->get_accValue(var_id, value);
+  }
+  return E_FAIL;
+}
+
+IFACEMETHODIMP AccessibilityRootNode::put_accValue(VARIANT var_id,
+                                                   BSTR new_value) {
+  IAccessible* target;
+  if ((target = GetTargetAndChildID(&var_id))) {
+    return target->put_accValue(var_id, new_value);
+  }
+  return E_FAIL;
+}
+
+IFACEMETHODIMP AccessibilityRootNode::get_accSelection(VARIANT* selected) {
+  selected->vt = VT_EMPTY;
+  return S_OK;
+}
+
+IFACEMETHODIMP AccessibilityRootNode::accSelect(LONG flagsSelect,
+                                                VARIANT var_id) {
+  IAccessible* target;
+  if ((target = GetTargetAndChildID(&var_id))) {
+    return target->accSelect(flagsSelect, var_id);
+  }
+  return E_FAIL;
+}
+
+IFACEMETHODIMP AccessibilityRootNode::get_accHelpTopic(BSTR* help_file,
+                                                       VARIANT var_id,
+                                                       LONG* topic_id) {
+  if (help_file) {
+    *help_file = nullptr;
+  }
+  if (topic_id) {
+    *topic_id = -1;
+  }
+  return E_NOTIMPL;
+}
+
+IFACEMETHODIMP AccessibilityRootNode::put_accName(VARIANT var_id,
+                                                  BSTR put_name) {
+  return E_NOTIMPL;
+}
+
+void AccessibilityRootNode::SetWindow(IAccessible* window) {
+  window_accessible_ = window;
+}
+
+AccessibilityAlert* AccessibilityRootNode::GetOrCreateAlert() {
+  if (!alert_accessible_) {
+    CComObject<AccessibilityAlert>* instance = nullptr;
+    HRESULT hr = CComObject<AccessibilityAlert>::CreateInstance(&instance);
```

Comment:
Should we use smart pointers instead? https://learn.microsoft.com/en-us/windows/win32/LearnWin32/com-coding-practices#com-smart-pointers

Also, I believe `instance` will remain `nullptr` if creating the instance fails. Perhaps the program should exit on failure?

Reply:
> ... is based on the code that existed previously for creating a COM object out of the native view ...

Could you link this code?

> I believe instance will remain `nullptr` if creating the instance fails. Perhaps the program should exit on failure?

What are your thoughts on this?

Reply:
AX Platform Node rather, not the window, see https://github.com/flutter/engine/blob/main/third_party/accessibility/ax/platform/ax_platform_node_win.cc#L257
We can exit on failure

Reply:
Re pointers, in a lot of other code we use the `Microsoft::WRL::ComPtr` smart pointer. e.g. here:
https://github.com/flutter/engine/blob/b1f1ec2bae460237675dcb74cee17a0193675093/shell/platform/windows/window.cc#L212-L217

Reply:
As far as I can tell, for the `ComPtr`, you need to have the actual instance to pass to the pointer (`root_view` in this case), and we cannot directly create `AccessibilityRootNode` by its constructor because it is an abstract subclass of a COM object. In this example, `root_view` holds the result of `GetNativeVeiwAccessible`, which, if it is an `AXPlatformNodeWin`, is created by a call to `CComObject::CreateInstance` in `AXPlatformNodeWin::Create`.